File: apps/coins/moarchy_coins/market.py
"""The market: what CoinGecko says, and what a phone can do with it.

Nothing here imports GTK, which is what makes the half of this app that can be
wrong -- parsing somebody else's JSON, and turning a number into the string a
person reads -- testable on any machine with a Python. The same split Habits
has between `habits.py` and its window, and for the same reason.

**The endpoint is `/coins/markets`, once, for the whole list.** One request
returns a hundred coins with their price, their rank, their cap and their day,
which is every number this app draws. The alternative -- `/simple/price` per
coin, or the per-coin endpoint for a detail page -- is a hundred requests for
the same screen, on a connection that is metered and a radio that is the second
biggest battery draw on the phone.

**No API key is required and one is used if offered.** The keyless tier is
rate-limited per IP at somewhere around five to fifteen calls a minute, shared
with everybody else behind the same carrier NAT, so a 429 is an ordinary
answer rather than an error -- `MarketError.retry_after` carries what to do
about it and the window backs off. `MOARCHY_COINS_KEY` is sent as
`x-cg-demo-api-key` for anybody who has registered one.

**Every number is optional.** A new listing has no rank, a coin that has not
traded today has no 24-hour change, and a delisted one has no price at all.
Each is dropped or drawn as a dash rather than being allowed to raise: a list
of a hundred coins that refuses to draw because the eighty-third has a null in
it is the failure mode to design against.
"""

from __future__ import annotations

import http.client
import json
import math
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass

API = "https://api.coingecko.com/api/v3/coins/markets"

# Who is asking. CoinGecko asks for a UA that identifies the client, and an app
# that names itself is one whose traffic can be recognised and blocked on its
# own rather than with every other keyless caller.
AGENT = "moarchy-coins/0.1.0 (+https://github.com/SimonSchubert/moarchy-apps)"

# Twelve seconds. A phone on a cell connection is slow rather than absent, and
# the usual ten is short enough to fail a request that would have arrived; the
# fetch is on a thread either way, so nothing is waiting on this but the word
# "Updating" in the header.
TIMEOUT = 12.0

# How many coins the list holds. A hundred is the length of every "top coins"
# list there has ever been, is about 80 kB of JSON, and is as far down the
# rankings as a person scrolls before they search instead. The endpoint's own
# ceiling is 250.
TOP = 100
PER_PAGE_MAX = 250

# The most a response may be before it is refused unread. 250 coins is a couple
# of hundred kilobytes; four megabytes is a body that has gone wrong, and on a
# metered connection an unbounded read is the expensive kind of wrong.
MAX_BYTES = 4 * 1024 * 1024

# The prices are in this unless something says otherwise. There is no picker in
# the app yet -- see the README -- so this and MOARCHY_COINS_CURRENCY are how it
# changes, and everything that formats a number takes the code rather than
# assuming the dollar.
CURRENCY = "usd"

# Currencies that have a sign worth drawing. Anything else is drawn as its code,
# which is how a currency nobody wrote a symbol for still reads correctly rather
# than silently becoming dollars.
SIGNS = {
    "usd": "$",
    "eur": "€",
    "gbp": "£",
    "jpy": "¥",
    "cny": "¥",
    "inr": "₹",
    "krw": "₩",
    "rub": "₽",
    "btc": "₿",
}

# What a 429 costs when the answer did not say how long to wait. CoinGecko's
# keyless limit is a handful of calls a minute per address, so a minute or two
# is the difference between being let back in and being refused again.
RATE_LIMIT_S = 120.0

# A number that is not there. An em dash rather than "0" or "n/a", because a
# missing 24-hour change and a flat one are different facts and a column of
# zeroes would be a lie told in the same typeface as the truth.
DASH = "—"

# Where a price stops being read as a price. Ten thousand is where the cents
# stop meaning anything -- nobody reads Bitcoin's -- and it is also where the
# string stops fitting the column: "$77,243" is seven characters and
# "$9,999.99" is nine, which is what a 360px row has room for beside a name.
# Below a hundredth the significant digits are all past the fourth place, so
# there the number of decimals is computed rather than fixed.
BIG = 10_000.0
SMALL = 0.01
FIGURES = 4
MAX_PLACES = 10

# Market caps, the way every exchange writes them.
UNITS = ((1e12, "T"), (1e9, "B"), (1e6, "M"), (1e3, "K"))


class MarketError(Exception):
    """Something a person can be told, and a hint about when to try again.

    Every failure in this module arrives as one of these carrying a sentence in
    plain words, because all of them end up in the same place: one line under a
    list of prices that are now older than they were. `retry_after` is seconds,
    and is 0 when the answer carried no such advice.
    """

    def __init__(self, message: str, *, retry_after: float = 0.0) -> None:
        super().__init__(message)
        self.retry_after = retry_after


@dataclass(frozen=True)
class Coin:
    """One row: everything drawn about a coin, and nothing else.

    Deliberately not the whole record. The endpoint returns thirty fields per
    coin -- supply, all-time highs, the date of each, a logo URL -- and this app
    draws seven of them. Keeping the other twenty-three would make the cache
    five times the size for a screen that never shows them, and would invite a
    detail page to be built out of whatever happened to be lying around rather
    than out of a request made for it.
    """

    id: str
    symbol: str
    name: str
    rank: int
    price: float
    change: float | None
    cap: float

    def matches(self, query: str) -> bool:
        """Is this the coin somebody is typing the name of?

        The front of the symbol, or the front of any word in the name. Not a
        substring anywhere: "itc" matching Bitcoin means a search box that fills
        with coincidences, and the coin somebody means is almost always one
        whose name or symbol starts that way. Any *word* rather than the first
        one, because "cash" has to find Bitcoin Cash and "inu" has to find Shiba
        Inu -- which is the same rule a launcher uses on app names.
        """
        text = query.strip().lower()
        if not text:
            return True
        if self.symbol.lower().startswith(text):
            return True
        return any(word.startswith(text) for word in self.name.lower().split())

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "symbol": self.symbol,
            "name": self.name,
            "rank": self.rank,
            "price": self.price,
            "change": self.change,
            "cap": self.cap,
        }

    @classmethod
    def from_dict(cls, data: dict) -> Coin:
        """One coin back out of our own cache file.

        Separate from `parse` below on purpose: that reads CoinGecko's shape and
        this reads ours. Folding them together would mean a change to the API's
        field names quietly rewriting what a cache written last week means.
        """
        if not isinstance(data, dict):
            raise TypeError("coin is not a table")
        price = _number(data.get("price"))
        if not data.get("id") or price is None:
            raise ValueError("coin has no id or no price")
        return cls(
            id=str(data["id"]),
            symbol=str(data.get("symbol") or ""),
            name=str(data.get("name") or data["id"]),
            rank=int(_number(data.get("rank")) or 0),
            price=price,
            change=_number(data.get("change")),
            cap=_number(data.get("cap")) or 0.0,
        )


def _number(value: object) -> float | None:
    """A JSON number, or None for anything that is not one.

    `isinstance(True, int)` is True in Python, and a bool where a price should
    be would otherwise become 1.0 and be drawn as a dollar.
    """
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return None
    number = float(value)
    return None if math.isnan(number) or math.isinf(number) else number


def parse(record: object, rank: int) -> Coin | None:
    """One coin out of CoinGecko's shape, or None if it is not usable.

    `rank` is where the coin sat in the answer, and stands in for a null
    `market_cap_rank` -- which a coin can genuinely have while still being the
    fortieth largest thing in a list ordered by market cap.
    """
    if not isinstance(record, dict):
        return None
    identifier = record.get("id")
    price = _number(record.get("current_price"))
    if not isinstance(identifier, str) or not identifier or price is None or price <= 0:
        # No price is not a coin worth a row: every column on the right of the
        # screen is derived from it.
        return None
    reported = _number(record.get("market_cap_rank"))
    return Coin(
        id=identifier,
        symbol=str(record.get("symbol") or "")[:8].upper(),
        name=str(record.get("name") or identifier),
        rank=int(reported) if reported and reported > 0 else rank,
        price=price,
        change=_number(record.get("price_change_percentage_24h")),
        cap=_number(record.get("market_cap")) or 0.0,
    )


def parse_markets(payload: object) -> list[Coin]:
    """The whole answer, with the unusable rows left out rather than fatal."""
    if not isinstance(payload, list):
        raise MarketError("CoinGecko sent something that is not a list of coins.")
    coins = []
    for position, record in enumerate(payload, start=1):
        coin = parse(record, position)
        if coin is not None:
            coins.append(coin)
    if not coins and payload:
        # Every row unusable is not "a quiet market": it is the shape of the
        # answer having changed, and saying so beats drawing an empty list.
        raise MarketError("CoinGecko sent coins in a shape this app cannot read.")
    return coins


# --- the wire ------------------------------------------------------------


def _retry_after(error: urllib.error.HTTPError) -> float:
    header = error.headers.get("Retry-After") if error.headers else None
    if isinstance(header, str) and header.strip().isdigit():
        return float(header.strip())
    return 0.0


def _get(url: str, *, key: str = "", timeout: float = TIMEOUT) -> object:
    """One GET, and a sentence for every way it can fail."""
    # The URL is always one this module built out of `API` and an urlencoded
    # query, so there is no scheme here a caller could have chosen.
    request = urllib.request.Request(
        url, headers={"User-Agent": AGENT, "Accept": "application/json"}
    )
    if key:
        request.add_header("x-cg-demo-api-key", key)
    try:
        with urllib.request.urlopen(request, timeout=timeout) as response:
            raw = response.read(MAX_BYTES + 1)
    except urllib.error.HTTPError as exc:
        if exc.code == 429:
            raise MarketError(
                "CoinGecko is rate-limiting this connection.",
                retry_after=_retry_after(exc) or RATE_LIMIT_S,
            ) from exc
        if 500 <= exc.code < 600:
            raise MarketError("CoinGecko is having trouble.") from exc
        raise MarketError(f"CoinGecko refused the request ({exc.code}).") from exc
    except (OSError, http.client.HTTPException) as exc:
        # URLError, a timeout, a reset, a half-sent body: on a phone these are
        # all one thing -- the radio -- and one sentence is the honest report.
        raise MarketError("No answer from CoinGecko.") from exc
    if len(raw) > MAX_BYTES:
        raise MarketError("CoinGecko sent more than this app will read.")
    try:
        return json.loads(raw)
    except ValueError as exc:
        raise MarketError("CoinGecko sent something that is not JSON.") from exc


class Live:
    """CoinGecko over HTTPS, which is the only thing in this app that is.

    An object rather than two functions so that the window can be handed
    something else entirely: the tests pass a stand-in with these two methods,
    and `MOARCHY_COINS_OFFLINE` passes nothing at all. The app has no idea which
    it has, which is why "the screenshots never touch the network" is a property
    of one environment variable rather than of a mode inside the window.
    """

    def __init__(
        self,
        *,
        currency: str = CURRENCY,
        count: int = TOP,
        key: str = "",
        timeout: float = TIMEOUT,
    ) -> None:
        self.currency = currency.lower() or CURRENCY
        self.count = max(1, min(count, PER_PAGE_MAX))
        self.key = key
        self.timeout = timeout

    def _url(self, **extra: str) -> str:
        query = {
            "vs_currency": self.currency,
            "order": "market_cap_desc",
            "per_page": str(self.count),
            "page": "1",
            # No sparkline. It is 168 hourly prices per coin -- seven times the
            # size of the whole rest of the answer -- for a chart this app does
            # not draw. See the README.
            "sparkline": "false",
            "price_change_percentage": "24h",
            "locale": "en",
        }
        query.update(extra)
        return f"{API}?{urllib.parse.urlencode(query)}"

    def markets(self) -> list[Coin]:
        """The top `count` coins by market capitalisation."""
        return parse_markets(_get(self._url(), key=self.key, timeout=self.timeout))

    def by_ids(self, ids: list[str]) -> list[Coin]:
        """Named coins, whatever their rank.

        The second request, and only when a starred coin has fallen out of the
        top hundred -- which is the one case where the list on screen cannot
        answer for a coin somebody asked to watch. An empty list of ids is not a
        request: without `ids` this endpoint answers with the whole market.
        """
        wanted = [i for i in ids if i][:PER_PAGE_MAX]
        if not wanted:
            return []
        url = self._url(ids=",".join(wanted), per_page=str(len(wanted)))
        return parse_markets(_get(url, key=self.key, timeout=self.timeout))


# --- numbers as somebody reads them --------------------------------------


def sign(currency: str) -> str:
    """The currency's mark, or its code and a space."""
    code = (currency or CURRENCY).lower()
    return SIGNS.get(code, f"{code.upper()} ")


def _places(value: float) -> str:
    """A price with as many decimals as it has significant digits, near zero.

    A coin priced at 0.0000132 has four figures worth reading and they all sit
    past the fourth decimal place, so a fixed two -- or a fixed eight -- is
    either a row of zeroes or a row of noise. This counts the leading zeroes and
    keeps four figures after them.
    """
    if value >= BIG:
        return f"{value:,.0f}"
    if value >= 1.0:
        return f"{value:,.2f}"
    if value >= SMALL:
        return f"{value:.4f}"
    leading = math.floor(-math.log10(value))
    return f"{value:.{min(leading + FIGURES, MAX_PLACES)}f}"


def money(value: float | None, currency: str = CURRENCY) -> str:
    if value is None or value <= 0:
        return DASH
    return f"{sign(currency)}{_places(value)}"


def compact(value: float | None, currency: str = CURRENCY) -> str:
    """A market cap: three significant figures and a letter.

    $1.55 T rather than $1,551,224,442,911, which is twenty characters of a
    360-pixel row spent on digits that change every second and mean nothing
    individually.
    """
    if value is None or value <= 0:
        return DASH
    for scale, suffix in UNITS:
        if value >= scale:
            scaled = value / scale
            if scaled < 10:
                return f"{sign(currency)}{scaled:.2f} {suffix}"
            if scaled < 100:
                return f"{sign(currency)}{scaled:.1f} {suffix}"
            return f"{sign(currency)}{scaled:.0f} {suffix}"
    return f"{sign(currency)}{value:,.0f}"


def percent(value: float | None) -> str:
    """The day, as a signed percentage.

    Always signed. The colour says up or down too, and the colour is the half a
    person sees first -- but roughly one man in twelve cannot tell this app's
    green from its red, and for them the sign *is* the answer.
    """
    if value is None:
        return DASH
    return f"{value:+.2f}%"


def direction(value: float | None) -> str:
    """ "up", "down" or "flat", for the class the figure is drawn in."""
    if value is None:
        return "flat"
    # Under half a hundredth rounds to +0.00%, and a plus sign in green over a
    # number that is not moving is a claim the figure beside it does not make.
    if value >= 0.005:
        return "up"
    if value <= -0.005:
        return "down"
    return "flat"


def freshness(seconds: float) -> str:
    """How old the prices are, in the words somebody would use.

    A clock that has gone backwards -- a phone that has just picked up NTP after
    being off for a week, which is the ordinary case rather than a strange one --
    reads as "just now" rather than as a negative number of minutes.
    """
    if seconds < 60:
        return "just now"
    if seconds < 3600:
        return f"{int(seconds // 60)} min ago"
    if seconds < 86400:
        hours = int(seconds // 3600)
        return f"{hours} hour ago" if hours == 1 else f"{hours} hours ago"
    days = int(seconds // 86400)
    return "yesterday" if days == 1 else f"{days} days ago"

File: plugins/org.moarchy.coins/Market.js
// The market: what CoinGecko says, and what a phone can do with it.
//
// The port of apps/coins/moarchy_coins/market.py, and the same split it makes:
// nothing here touches QML, so the half of this app that can be wrong --
// parsing somebody else's JSON, and turning a number into the string a person
// reads -- is testable with no display, exactly as the Python half was.
//
// The numbers are the interesting part and they are ported literally rather
// than approximated: a price is grouped and rounded the same way, a cap gets
// the same three significant figures, and a change always carries its sign
// because roughly one man in twelve cannot tell this app's green from its red.
.pragma library

var API = "https://api.coingecko.com/api/v3/coins/markets"

// Who is asking. An app that names itself is one whose traffic can be
// recognised and blocked on its own rather than with every other keyless
// caller.
var AGENT = "moarchy-coins/0.1.1 (+https://github.com/SimonSchubert/moarchy-apps)"

var TIMEOUT = 12
var TOP = 100
var PER_PAGE_MAX = 250
var CURRENCY = "usd"

// CoinGecko's keyless limit is a handful of calls a minute per address, so a
// minute or two is the difference between being let back in and being refused
// again.
var RATE_LIMIT_S = 120

// A minute is the interval a coin tracker is worth having at.
var REFRESH_S = 60

// Larger than any list this app asks for, so an unranked coin sorts to the end
// rather than into the middle, and small enough to stay three digits in the
// disc it is drawn in.
var UNRANKED = 999

// An em dash rather than "0" or "n/a": a missing 24-hour change and a flat one
// are different facts, and a column of zeroes would be a lie told in the same
// typeface as the truth.
var DASH = "—"

var BIG = 10000
var SMALL = 0.01
var FIGURES = 4
var MAX_PLACES = 10

var SIGNS = {
  usd: "$", eur: "€", gbp: "£", jpy: "¥", cny: "¥",
  inr: "₹", krw: "₩", rub: "₽", btc: "₿"
}

var UNITS = [[1e12, "T"], [1e9, "B"], [1e6, "M"], [1e3, "K"]]

// --- numbers as somebody reads them --------------------------------------

// Python's format spec does the thousands separator with `,`; V4 has no such
// spec and Intl is not worth depending on inside a shell, so the grouping is
// done by hand on the integer side only.
function group(digits) {
  var out = ""
  var n = String(digits)
  for (var i = 0; i < n.length; i++) {
    if (i > 0 && (n.length - i) % 3 === 0) out += ","
    out += n[i]
  }
  return out
}

function fixed(value, places) {
  var s = value.toFixed(places)
  var dot = s.indexOf(".")
  var whole = dot < 0 ? s : s.slice(0, dot)
  var rest = dot < 0 ? "" : s.slice(dot)
  return group(whole) + rest
}

function sign(currency) {
  var code = String(currency || CURRENCY).toLowerCase()
  return SIGNS[code] !== undefined ? SIGNS[code] : code.toUpperCase() + " "
}

// A price with as many decimals as it has significant digits, near zero. A coin
// at 0.0000132 has four figures worth reading and they all sit past the fourth
// decimal place, so a fixed two -- or a fixed eight -- is either a row of
// zeroes or a row of noise.
function places(value) {
  if (value >= BIG) return fixed(value, 0)
  if (value >= 1) return fixed(value, 2)
  if (value >= SMALL) return value.toFixed(4)
  var leading = Math.floor(-Math.log(value) / Math.LN10)
  return value.toFixed(Math.min(leading + FIGURES, MAX_PLACES))
}

function money(value, currency) {
  if (value === null || value === undefined || !(value > 0)) return DASH
  return sign(currency) + places(value)
}

// A market cap: three significant figures and a letter. $1.55 T rather than
// $1,551,224,442,911, which is twenty characters of a 360px row spent on digits
// that change every second and mean nothing individually.
function compact(value, currency) {
  if (value === null || value === undefined || !(value > 0)) return DASH
  for (var i = 0; i < UNITS.length; i++) {
    var scale = UNITS[i][0], suffix = UNITS[i][1]
    if (value >= scale) {
      var scaled = value / scale
      var dp = scaled < 10 ? 2 : (scaled < 100 ? 1 : 0)
      return sign(currency) + scaled.toFixed(dp) + " " + suffix
    }
  }
  return sign(currency) + fixed(value, 0)
}

// Always signed. The colour says up or down too, and the colour is the half a
// person sees first -- but for anyone who cannot tell them apart the sign is
// the answer.
function percent(value) {
  if (value === null || value === undefined) return DASH
  var s = Math.abs(value).toFixed(2)
  return (value < 0 ? "-" : "+") + s + "%"
}

function direction(value) {
  if (value === null || value === undefined) return "flat"
  // Under half a hundredth rounds to +0.00%, and a plus sign in green over a
  // number that is not moving is a claim the figure beside it does not make.
  if (value >= 0.005) return "up"
  if (value <= -0.005) return "down"
  return "flat"
}

// A clock that has gone backwards -- a phone that has just picked up NTP after
// being off for a week -- reads as "just now" rather than as a negative number.
function freshness(seconds) {
  if (seconds < 60) return "just now"
  if (seconds < 3600) return Math.floor(seconds / 60) + " min ago"
  if (seconds < 86400) {
    var hours = Math.floor(seconds / 3600)
    return hours === 1 ? "1 hour ago" : hours + " hours ago"
  }
  var days = Math.floor(seconds / 86400)
  return days === 1 ? "yesterday" : days + " days ago"
}

// --- CoinGecko's shape ----------------------------------------------------

// `typeof true === "boolean"`, so a bool cannot arrive here as 1 and be drawn
// as a dollar the way Python's isinstance(True, int) would have allowed.
function number(value) {
  if (typeof value !== "number" || !isFinite(value)) return null
  return value
}

// In an ordered answer the rank is where the coin sat in it, and
// `market_cap_rank` is ignored: that field disagrees with the ordering it
// arrives in, and drawn faithfully it gives a list numbered 8, 9, 9, 10, which
// reads as a broken app rather than as a quirk of somebody else's field.
function parse(record, position, ordered) {
  if (!record || typeof record !== "object") return null
  var id = record.id
  var price = number(record.current_price)
  if (typeof id !== "string" || !id || price === null || price <= 0) return null
  var rank
  if (ordered === false) {
    var reported = number(record.market_cap_rank)
    rank = reported && reported > 0 ? Math.round(reported) : UNRANKED
  } else {
    rank = position
  }
  return {
    id: id,
    symbol: String(record.symbol || "").slice(0, 8).toUpperCase(),
    name: String(record.name || id),
    rank: rank,
    price: price,
    change: number(record.price_change_percentage_24h),
    cap: number(record.market_cap) || 0,
    image: iconUrl(record.image)
  }
}

function parseMarkets(text, ordered) {
  var payload
  try {
    payload = JSON.parse(String(text || ""))
  } catch (e) {
    return { error: "CoinGecko sent something this app cannot read.", coins: [] }
  }
  if (!payload || payload.constructor !== Array)
    return { error: "CoinGecko sent something that is not a list of coins.", coins: [] }
  var coins = []
  for (var i = 0; i < payload.length; i++) {
    var coin = parse(payload[i], i + 1, ordered)
    if (coin) coins.push(coin)
  }
  // Every row unusable is not "a quiet market": it is the shape of the answer
  // having changed, and saying so beats drawing an empty list.
  if (!coins.length && payload.length)
    return { error: "CoinGecko sent coins in a shape this app cannot read.", coins: [] }
  return { error: "", coins: coins }
}

function url(currency, count, ids) {
  var per = Math.max(1, Math.min(count || TOP, PER_PAGE_MAX))
  var q = "vs_currency=" + encodeURIComponent(String(currency || CURRENCY).toLowerCase())
        + "&order=market_cap_desc"
        + "&per_page=" + per
        + "&page=1"
        // No sparkline: 168 hourly prices per coin, seven times the size of the
        // whole rest of the answer, for a chart this app does not draw.
        + "&sparkline=false"
        + "&price_change_percentage=24h"
        + "&locale=en"
  if (ids && ids.length) q += "&ids=" + encodeURIComponent(ids.join(","))
  return API + "?" + q
}

// The front of the symbol, or the front of any word in the name. Not a
// substring anywhere: "itc" matching Bitcoin means a search box that fills with
// coincidences. Any *word*, because "cash" has to find Bitcoin Cash and "inu"
// has to find Shiba Inu.
function matches(coin, query) {
  var text = String(query || "").trim().toLowerCase()
  if (!text) return true
  if (String(coin.symbol || "").toLowerCase().indexOf(text) === 0) return true
  var words = String(coin.name || "").toLowerCase().split(/\s+/)
  for (var i = 0; i < words.length; i++)
    if (words[i].indexOf(text) === 0) return true
  return false
}

function filter(coins, query) {
  var out = []
  for (var i = 0; i < coins.length; i++)
    if (matches(coins[i], query)) out.push(coins[i])
  return out
}

// The starred coins, in the order they were starred -- not in rank order, which
// is the obvious alternative and is wrong on a phone: a watchlist sorted by
// market cap reorders itself under a thumb halfway down it.
//
// A star with no price behind it is left out rather than drawn as a row of
// dashes, which lasts exactly as long as it takes the next fetch to ask for it
// by name.
function starred(coins, favourites, query) {
  var known = {}
  for (var i = 0; i < coins.length; i++) known[coins[i].id] = coins[i]
  var out = []
  for (var j = 0; j < favourites.length; j++) {
    var coin = known[favourites[j]]
    if (coin && matches(coin, query)) out.push(coin)
  }
  return out
}

// Starred coins the last answer did not cover, which is what the second request
// exists for: a coin can only be starred from the top hundred, but it can fall
// out of it afterwards, and a watchlist quietly showing last week's price is
// worse than one that says it cannot.
function missing(coins, favourites) {
  var known = {}
  for (var i = 0; i < coins.length; i++) known[coins[i].id] = true
  var out = []
  for (var j = 0; j < favourites.length; j++)
    if (!known[favourites[j]]) out.push(favourites[j])
  return out
}

// --- the coin's own logo --------------------------------------------------

// theme.py argued against logos on three grounds: a hundred requests to a CDN,
// each one telling that CDN which coins somebody watches, for pictures 24px
// wide. Two of the three have answers and the measurements are why:
//
//   * The URL arrives in the market answer already, so finding it costs no
//     request at all -- market.py simply dropped the field.
//   * `small` is 50x50 and about 2.5 kB, so the whole top hundred is ~245 kB
//     once and then never again. `large`, which is what the API hands over, is
//     12 kB each and 1.2 MB the set -- which is the version of this that
//     deserved to be refused.
//   * Every coin in the list is fetched, not the starred ones: asking for all
//     hundred says exactly what asking for the top hundred already said, while
//     asking for four would name the four. The privacy objection argues for
//     fetching *more*, not less.
//
// What stays true is the last one: at 36 points a logo is a smudge for many
// coins. So the disc is not replaced, it is what shows until a file is there --
// and on a phone that has never been online, it is all there ever is.
var ICON_SIZE = "small"

// The API hands over the 250px one. The path segment is the size, so this is a
// rewrite rather than a second request to discover a smaller variant.
function iconUrl(url) {
  var s = String(url || "")
  if (!s) return ""
  if (s.indexOf("/large/") < 0) return s
  return s.replace("/large/", "/" + ICON_SIZE + "/")
}

// The file a coin's logo is cached at. Keyed by id and not by the URL's own
// name, because the id is what the row already has and what the cache file
// already carries -- and because two coins can ship a `logo.png`.
function iconFile(coinId) {
  return String(coinId).replace(/[^A-Za-z0-9._-]/g, "_") + ".png"
}

// --- the disc a coin wears ------------------------------------------------

// The eight hue roles a theme names, as the eight discs a coin can wear. All of
// them rather than a chosen few: with a hundred rows on the page the point is
// that two coins next to each other are unlikely to match, and dropping the
// awkward hues would make that likelier rather than the list prettier.
var BADGE_HUES = ["blue", "cyan", "magenta", "orange", "green", "yellow", "brown", "red"]

// How much of the hue goes into the disc. Solid enough to tell eight of them
// apart at arm's length, pale enough that the theme's own text colour reads on
// top of it in both a light theme and a dark one.
var BADGE_TINT = 0.28

var CRC_TABLE = null

function crcTable() {
  if (CRC_TABLE) return CRC_TABLE
  var table = []
  for (var n = 0; n < 256; n++) {
    var c = n
    for (var k = 0; k < 8; k++) c = (c & 1) ? (0xEDB88320 ^ (c >>> 1)) : (c >>> 1)
    table[n] = c >>> 0
  }
  CRC_TABLE = table
  return table
}

// zlib.crc32 of the UTF-8 bytes, which is what theme.py uses -- and it uses it
// rather than a hash because Python's hash() is salted per process and would
// give a coin a different colour on every launch, turning the one property this
// is for ("my coin is the blue one") into noise. Ported rather than replaced so
// a phone with both halves installed draws one coin one colour.
function crc32(text) {
  var table = crcTable()
  var crc = 0xFFFFFFFF
  var s = String(text)
  for (var i = 0; i < s.length; i++) {
    var code = s.charCodeAt(i)
    // UTF-8, by hand: the ids are ASCII today, but "shiba-inu" is not a promise
    // and a multibyte id must hash to what Python's .encode("utf-8") hashes to.
    var bytes
    if (code < 0x80) bytes = [code]
    else if (code < 0x800) bytes = [0xC0 | (code >> 6), 0x80 | (code & 0x3F)]
    else bytes = [0xE0 | (code >> 12), 0x80 | ((code >> 6) & 0x3F), 0x80 | (code & 0x3F)]
    for (var b = 0; b < bytes.length; b++)
      crc = table[(crc ^ bytes[b]) & 0xFF] ^ (crc >>> 8)
  }
  return (crc ^ 0xFFFFFFFF) >>> 0
}

function badgeHue(coinId) {
  return BADGE_HUES[crc32(coinId) % BADGE_HUES.length]
}
